File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/weaveit/seatingplacesuggestions/SeatingPlace.java
package org.weaveit.seatingplacesuggestions;

public class SeatingPlace {

    private final String rowName;
    private final int number;
    private final PricingCategory pricingCategory;
    private SeatingPlaceAvailability seatingPlaceAvailability;

    public SeatingPlace(String rowName, int number, PricingCategory pricingCategory, SeatingPlaceAvailability seatingPlaceAvailability) {
        this.rowName = rowName;
        this.number = number;
        this.pricingCategory = pricingCategory;
        this.seatingPlaceAvailability = seatingPlaceAvailability;
    }
    public boolean isAvailable() {
        return seatingPlaceAvailability == SeatingPlaceAvailability.AVAILABLE;
    }

    public boolean matchCategory(PricingCategory pricingCategory) {
        return this.pricingCategory == pricingCategory;
    }

    public void allocate() {
        if (seatingPlaceAvailability == SeatingPlaceAvailability.AVAILABLE)
            seatingPlaceAvailability = SeatingPlaceAvailability.ALLOCATED;
    }

    @Override
    public String toString() {
        return rowName + number;
    }
}

File: TheaterSuggestions/CSharp/ScreeningSeatingArrangementSuggestions/SeatingPlace.cs
namespace SeatsSuggestions;

public class SeatingPlace : Value.ValueType<SeatingPlace>
{
    public SeatingPlace(string rowName, int number, PricingCategory pricingCategory,
        SeatingPlaceAvailability seatingPlaceAvailability)
    {
        RowName = rowName;
        Number = number;
        PricingCategory = pricingCategory;
        SeatingPlaceAvailability = seatingPlaceAvailability;
    }

    public string RowName { get; }
    public int Number { get; }
    public PricingCategory PricingCategory { get; }
    public SeatingPlaceAvailability SeatingPlaceAvailability { get; }

    public bool IsAvailable()
    {
        return SeatingPlaceAvailability == SeatingPlaceAvailability.Available;
    }

    public override string ToString()
    {
        return $"{RowName}{Number}";
    }

    public bool MatchCategory(PricingCategory pricingCategory)
    {
        if (pricingCategory == PricingCategory.Ignore) return true;
        return PricingCategory == pricingCategory;
    }

    public SeatingPlace Allocate()
    {
        if (SeatingPlaceAvailability == SeatingPlaceAvailability.Available)
            return new(RowName, Number, PricingCategory, SeatingPlaceAvailability.Allocated);
        return this;
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return new object[] { RowName, Number, PricingCategory, SeatingPlaceAvailability };
    }
}